apollo-router/src/services/router.rs

```diff
@@ -0,0 +1,145 @@
+pub type ServiceResult = Result<Response, BoxError>;
+pub type Body = hyper::Body;
+
+assert_impl_all!(Request: Send);
+/// Represents the router processing step of the processing pipeline.
+///
+/// This consists of the parsed graphql Request, HTTP headers and contextual data for extensions.
+#[non_exhaustive]
+pub struct Request {
+    /// Original request to the Router.
+    pub router_request: http::Request<Body>,
+
+    /// Context for extension
+    pub context: Context,
+}
+
+impl From<http::Request<Body>> for Request {
+    fn from(router_request: http::Request<Body>) -> Self {
+        Self {
+            router_request,
+            context: Context::new(),
+        }
+    }
+}
+
+use displaydoc::Display;
+use thiserror::Error;
+
+#[derive(Error, Display, Debug)]
+pub enum ParseError {
+    /// couldn't create a valid http GET uri '{0}'
+    InvalidUri(http::uri::InvalidUri),
+    /// couldn't urlencode the GraphQL request body '{0}'
+    UrlEncodeError(serde_urlencoded::ser::Error),
+    /// couldn't serialize the GraphQL request body '{0}'
+    SerializationError(serde_json::Error),
+}
+
+/// This is handy for tests.
+impl TryFrom<supergraph::Request> for Request {
+    type Error = ParseError;
+    fn try_from(request: supergraph::Request) -> Result<Self, Self::Error> {
+        let supergraph::Request {
+            context,
+            supergraph_request,
+        } = request;
+
+        let (mut parts, request) = supergraph_request.into_parts();
+
+        let router_request = if parts.method == Method::GET {
+            // get request
+            let get_path = serde_urlencoded::to_string(&[
+                ("query", request.query),
+                ("operationName", request.operation_name),
+                (
+                    "extensions",
+                    serde_json::to_string(&request.extensions).ok(),
+                ),
+                ("variables", serde_json::to_string(&request.variables).ok()),
+            ])
+            .map_err(ParseError::UrlEncodeError)?;
+
+            parts.uri = format!("{}?{}", parts.uri, get_path)
+                .parse()
+                .map_err(ParseError::InvalidUri)?;
+
+            http::Request::from_parts(parts, Body::empty())
+        } else {
+            http::Request::from_parts(
+                parts,
+                Body::from(serde_json::to_vec(&request).map_err(ParseError::SerializationError)?),
+            )
+        };
+        Ok(Self {
+            router_request,
+            context,
+        })
+    }
+}
+
+assert_impl_all!(Response: Send);
+#[non_exhaustive]
+pub struct Response {
+    pub response: http::Response<Body>,
+    pub context: Context,
+}
+
+impl From<http::Response<Body>> for Response {
+    fn from(response: http::Response<Body>) -> Self {
+        Self {
+            response,
+            context: Context::new(),
+        }
+    }
+}
+
+impl Response {
+    pub async fn next_response(&mut self) -> Option<Result<Bytes, hyper::Error>> {
```

Comment:
As we have a type alias for `Body` then also consider having one for `Error`

Reply:
fixed in [64bf759](https://github.com/apollographql/router/pull/2170/commits/64bf7593f450e2359a03d76f8b432c2f4c28f13b)